### PR TITLE
minimize thread-stalling np.random.randint() calls

### PR DESCRIPTION
**Pull request by gojomo**
Probable fix for #336 - something about the numpy randint() (perhaps its own use of NOGIL) makes calling it once per word (to generate `reduced_windows`) wreak havoc on multiple worker processing.

### CODE REVIEW

File: gensim/models/word2vec_inner.pyx [hunk 1]
```diff
@@ -633,6 +632,9 @@ def train_sentence_sg(model, sentence, alpha, _work):
             else:
                 codelens[i] = 1
             result += 1
+    # single randint() call avoids a big thread-sync slowdown
+    for i, item in enumerate(np.random.randint(0, window, sentence_len)):
+        reduced_windows[i] = item
```
Comment (piskvorky) [inline gensim/models/word2vec_inner.pyx]: Why not just assign the result of `randint(low, high, size)` to `reduced_windows[:size]`? Is there a problem with that? 

Should be faster too.

Comment (gojomo) [inline gensim/models/word2vec_inner.pyx]: I hadn't even thought of that, being in C-thinking mode when I wrote it (and filling the `cdef np.uint32_t reduced_windows[MAX_SENTENCE_LEN]` array). But I just tried…

`reduced_windows[:sentence_len] = np.random.randint(0, window, sentence_len)`

…and got a couple cython errors. On the other hand, this variant…

`reduced_windows[:sentence_len] = <np.uint32_t *>(np.PyArray_DATA(np.random.randint(0, window, sentence_len)))`

…will compile, so I'm giving it a try now.

Comment (piskvorky) [inline gensim/models/word2vec_inner.pyx]: If it's too hairy or obscure then never mind. I haven't actually tried / profiled that change, it's just a minor nitpick.

**Comment by gojomo** [comment]
Casting the return ndarray from randint to be the new `reduced_windows` is sufficient and eliminates the loop/copy. (The stack-local preallocation for `reduced_windows` goes away too.)

**Comment by piskvorky** [comment]
Hmm, and when does the array get de-allocated? Is that handled by cython automagically, somehow?

**Comment by gojomo** [comment]
Hmm; the ndarray returned from randint() is a python object, so automatically managed as usual. But I don't know if the reference to the PyArray_DATA value is enough to keep that python object alive... so if GC (and reallocation) is possible before the cython function returns, there might be a risk of the array being clobbered by other uses. I don't yet know enough about CPython GC to know if that's a risk and a few searches haven't yet turned up a sure answer; I'll take a closer look tomorrow.

**Comment by piskvorky** [comment]
In that case, I propose to keep the simple loop for now :)

I'd really like us to focus on finalizing the doc2vec PR.

**Comment by gojomo** [comment]
OK, rolled back other variant, simple approach that's now in this branch/PR should be zero-risk/all-benefits, if anyone's eager to test.

**Comment by piskvorky** [comment]
Merging, thanks @gojomo !

**Pull request merged**